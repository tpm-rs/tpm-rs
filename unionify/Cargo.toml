[package]
name = "unionify"
version = "0.1.0"
edition = "2021"

[dependencies]
# unionify-derive is not part of workspace because we do not want anyone
# to accidentally import it
unionify-derive = { path = "./unionify-derive/" }

[dev-dependencies]
trybuild = { workspace = true }
