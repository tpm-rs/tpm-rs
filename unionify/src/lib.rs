//! [UnionSize] is a derivable trait that calculates the size of
//! `repr(C)` union part in a tagged or untagged enum.
//!
//! # Example
//! ```rust
//! use tpm2_rs_unionify::UnionSize;
//! #[derive(UnionSize)]
//! enum Foo {
//!     V1(u8),
//!     V2([u32; 128]),
//!     V3{
//!         buffer: [u8;32]
//!     },
//!     V4
//! }
//! assert_eq!(Foo::UNION_SIZE, 128*32/8);
//! ```

pub use tpm2_rs_unionify_derive::UnionSize;

pub trait UnionSize {
    /// The size (in bytes) of `repr(C)`-union-equivalent to the algebraic enum in question.
    const UNION_SIZE: usize;
}

#[cfg(test)]

mod test {
    #[test]
    fn test_unionify() {
        let t = trybuild::TestCases::new();
        t.pass("tests/pass/*.rs");
        t.compile_fail("tests/fail/*.rs")
    }
}
