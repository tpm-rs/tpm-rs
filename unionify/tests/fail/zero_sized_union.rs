use unionify::UnionSize;

#[derive(UnionSize)]
pub enum Foo {
    A,
    B,
}

fn main() {}
