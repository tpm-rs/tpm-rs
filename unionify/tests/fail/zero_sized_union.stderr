error: unions cannot have zero fields
 --> tests/fail/zero_sized_union.rs:3:10
  |
3 | #[derive(UnionSize)]
  |          ^^^^^^^^^
  |
  = note: this error originates in the derive macro `UnionSize` (in Nightly builds, run with -Z macro-backtrace for more info)
